### PR TITLE
ENH: Add Batch context.

### PR DESCRIPTION
**Pull request by danielballan**

### CODE REVIEW

File: caproto/threading/client.py [hunk 1]
```diff
@@ -1789,6 +1792,110 @@ def __del__(self):
             pass
 
 
+class Batch:
```
Comment (klauer) [inline caproto/threading/client.py]: This `Batch` syntax/functionality is even nicer than I had imagined - nice work!

File: caproto/threading/client.py [hunk 2]
```diff
@@ -1789,6 +1792,110 @@ def __del__(self):
...
+    """
+    Accumulate requests and then issue them all in batch.
+
+    Parameters
+    ----------
+    timeout : number or None
+        Number of seconds to wait before ignoring late responses. Default
+        is 2.
```
Comment (klauer) [inline caproto/threading/client.py]: Maybe in this case a linear function, proportional to number of requests, would be reasonable?

File: caproto/threading/client.py [hunk 3]
```diff
@@ -1789,6 +1792,110 @@ def __del__(self):
...
+
+    Examples
+    --------
+    Read some PVs in batch and stash the readings in a dictionary as they
+    come in.
+
+    >>> readings = {}
+
+    >>> results = {}
+    >>> stash_result = lambda name, response: results.update({name: response.data})
```
Comment (klauer) [inline caproto/threading/client.py]: flake8: no lambda! 😄 
```
def stash_result(name, response):
    results[name] = response.data
```

^ seems a bit more clear for a basic example, I think.

File: caproto/threading/client.py [hunk 4]
```diff
@@ -1789,6 +1792,110 @@ def __del__(self):
...
+    timeout : number or None
+        Number of seconds to wait before ignoring late responses. Default
+        is 2.
+
+    Examples
+    --------
+    Read some PVs in batch and stash the readings in a dictionary as they
+    come in.
+
+    >>> readings = {}
```
Comment (klauer) [inline caproto/threading/client.py]: `readings` unused

File: caproto/threading/client.py [hunk 5]
```diff
@@ -1789,6 +1792,110 @@ def __del__(self):
...
+
+    >>> results = {}
+    >>> stash_result = lambda name, response: results.update({name: response.data})
+
+    >>> with Batch() as b:
+    ...     for pv in pvs:
+    ...         b.read(pv, functools.partial(stash_result, pv.name))
```
Comment (klauer) [inline caproto/threading/client.py]: I think it's good as-is, but thinking of alternatives: what if `read` also gave you a `Future` of sorts for getting the result of each request? A dictionary? Hmm...

File: caproto/threading/client.py (hunk 1 above)
Comment (danielballan) [reply caproto/threading/client.py]: Good suggestion of syntax!

File: caproto/threading/client.py (hunk 5 above)
Comment (danielballan) [reply caproto/threading/client.py]: I think the asyncio variety should return a Future. Trio doesn't _have_ futures, so not sure about that. I think we should add _something_ once we see how this shakes out in other frameworks.

File: caproto/threading/client.py (hunk 2 above)
Comment (danielballan) [reply caproto/threading/client.py]: Better do this on all the methods then. Punting to https://github.com/NSLS-II/caproto/issues/284 (but plan to address today)

**Comment by danielballan** [comment]
Tests passed; failed on flake8 again. Pushed fix to test code, reran tests locally to verify I didn't break the tests I touched. Merging.

**Pull request merged**